src/pybind/mgr/cephadm/agent.py

```diff
@@ -0,0 +1,218 @@
+        except Exception as e:
+            self.mgr.log.warning(f'Failed to handle update from agent: {e}')
+
+
+class AgentMessageThread(threading.Thread):
+    def __init__(self, host: str, port: int, counter: int, mgr: "CephadmOrchestrator") -> None:
+        self.mgr = mgr
+        self.host = host
+        self.port = port
+        self.counter = counter
+        super(AgentMessageThread, self).__init__(target=self.run)
+
+    def run(self) -> None:
+        for retry_wait in [3, 5]:
+            try:
+                agent_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                agent_socket.connect((self.mgr.inventory.get_addr(self.host), self.port))
+                agent_socket.send(str(self.counter).encode())
+                agent_response = agent_socket.recv(1024).decode()
+                self.mgr.log.debug(f'Received {agent_response} from agent on host {self.host}')
+                return
+            except ConnectionError as e:
+                # if it's a connection error, possibly try to connect again.
+                # We could have just deployed agent and it might not be ready
+                self.mgr.log.debug(
+                    f'Retrying connection to agent on {self.host} in {str(retry_wait)} seconds. Connection failed with: {e}')
+                time.sleep(retry_wait)
+            except Exception as e:
+                # if it's not a connection error, something has gone wrong. Give up.
+                self.mgr.log.debug(f'Failed to contact agent on host {self.host}: {e}')
+                return
+        return
+
+
+class CephadmAgentHelpers:
+    def __init__(self, mgr: "CephadmOrchestrator"):
+        self.mgr: "CephadmOrchestrator" = mgr
+
+    def _request_agent_acks(self, hosts: Set[str]) -> None:
+        for host in hosts:
+            self.mgr.cache.metadata_up_to_date[host] = False
+            if host not in self.mgr.cache.agent_counter:
+                self.mgr.cache.agent_counter[host] = 1
+            else:
+                self.mgr.cache.agent_counter[host] = self.mgr.cache.agent_counter[host] + 1
+            message_thread = AgentMessageThread(
+                host, self.mgr.agent_port, self.mgr.cache.agent_counter[host], self.mgr)
+            message_thread.start()
+
+    def _use_agent_metadata(self, host: str) -> bool:
+        # we're not using agents
+        if not self.mgr.use_agent:
+            return False
+        # no entry for agent in timestamp, it likely hasn't been deployed
+        if host not in self.mgr.cache.agent_timestamp:
+            return False
+        # agent hasn't reported in 2.5 * it's refresh rate. Something is likely wrong with it.
+        time_diff = self.mgr.cache.agent_timestamp[host] - datetime_now()
+        if time_diff.total_seconds() > 2.5 * float(self.mgr.agent_refresh_rate):
+            return False
+        return True
+
+    def _check_agents_ip_port(self) -> None:
```

Comment:
This can be dropped entirely if you return the mgr ip + port in the config deps from generate_config.